### PR TITLE
#169257410 Builds LineManager update request to appproved or declined

### PR DESCRIPTION
**Pull request by b0nbon1**
#### What does this PR do?
Line managers should be to approve or decline their direct report users' requests that are open.
#### Description of Task to be completed?
- [x] check the request, if the line manager can approve
- [x] check if the status requested to update exists
- [x]  check if the request has already been approved or declined
- [x] approve or decline the request successful
#### How should this be manually tested?
- git clone this branch
- run `npm install` to install dependencies 
- run `npm run migrate:undo` and `npm run migrate` to build the tables
- run `npm run watch` to start the server
- run signup two users with the data on the `POST {{url}}/api/v1/auth/signup`
- update one user to be a manager by using `Patch {{url}}/api/v1/auth/user/role`
- update the user using the line manager id `Patch {{url}}/api/v1/user/update-profile`
- using the user token create a trip
- using line manager token get the trips using endpoint `Patch {{url}}/api/v1/request/:id` and use `{ "status": "approved" }` to approve or `{ "status": "declined" }` to decline
#### Any background context you want to provide?
- set the line manager on the user before approving
- [implementation plan](https://docs.google.com/document/d/1M81hCB_qAQhr37Dt_K90FHjeO6v5FCQD9jRyvdS1zvQ/edit?usp=sharing)
#### What are the relevant pivotal tracker stories?
[#169257410](https://www.pivotaltracker.com/n/projects/2407417/stories/169257410)

#### Questions:
- Should the status to be update be passed on the `req.query` or `req.body`?

### CODE REVIEW

**Review by gildniy — DISMISSED** [review]
Looks great, I think it can be merged!

**Review by RaymondMwaura — COMMENTED** [review]
The functionality works as described. Great job! However, I think you should add some flexibility to it so that the request status can be switched between 'open', 'approved' and 'declined' at the manager's discretion. Currently, once the manager approves or declines a request, he can no longer reopen it.

File: src/services/request.service.js
Comment (collinewait) [inline src/services/request.service.js]: `approved` and `declined` can be placed in an `enum` like structure.

File: src/services/request.service.js
Comment (collinewait) [inline src/services/request.service.js]: It is hard to understand the `condition` in the if statement. The condition can be given a better name. A readable variable can take up `request[0] === 0`.

File: src/tests/approveRequest.test.js
Comment (collinewait) [inline src/tests/approveRequest.test.js]: constant `prefix` can be uppercase.

File: src/tests/approveRequest.test.js
Comment (collinewait) [inline src/tests/approveRequest.test.js]: `No access to the` looks to be incomplete. Same issue in the tests below.

File: src/middlewares/checkRequestLineManager.js
Comment (nshaphan) [inline src/middlewares/checkRequestLineManager.js]: You can move this to the request service, this block of code can be reused

**Review by nshaphan — DISMISSED** [review]

**Review by RaymondMwaura — DISMISSED** [review]
This is well implemented. Keep up the pace.

File: src/services/request.service.js
Comment (b0nbon1) [reply src/services/request.service.js]: Thanks for this. Yeah, I have placed them in an `array` then on `if` I used `array.includes()` function to check if it is the status. What do you think

File: src/services/request.service.js
Comment (b0nbon1) [reply src/services/request.service.js]: Thanks, working on it

File: src/services/request.service.js
Comment (collinewait) [reply src/services/request.service.js]: Reading from an array is very expensive. You can use an object which will give us a time complexity of O(1)

Comment (b0nbon1) [reply src/services/request.service.js]: nice suggestion, I didn't think about that. An object can work

**Review by gildniy — DISMISSED** [review]
LGTM after testing it, I think it can be merged

**Review by Mastel22 — DISMISSED** [review]
Tested and it's working fine. Nice job

File: src/services/request.service.js
Comment (collinewait) [inline src/services/request.service.js]: uppercase the keys. The object can also be outside the function.

**Review by RaymondMwaura — DISMISSED** [review]
This feature was swiftly and accurately implemented. Great job.

**Review by RaymondMwaura — APPROVED** [review]
This feature was swiftly and accurately implemented. Great job.

**Review by nshaphan — APPROVED** [review]
Great work @b0nbon1

**Review by collinewait — APPROVED** [review]

**Pull request merged**
